src/common/utils/getLeaderBoardDataFromGithub.ts

```diff
@@ -0,0 +1,134 @@
+import { z } from 'zod';
+
+const githubWeekStatesSchema = z.object({
+  w: z.number(),
+  a: z.number(),
+  d: z.number(),
+  c: z.number(),
+});
+
+const gitHubauthorSchema = z.object({
+  login: z.string(),
+  id: z.number(),
+  node_id: z.string(),
+  avatar_url: z.string(),
+  // ... (other author properties)
+});
+
+const gitHubContributorStatsSchema = z.array(
+  z.object({
+    total: z.number(),
+    weeks: z.array(githubWeekStatesSchema),
+    author: gitHubauthorSchema,
+  }),
+);
+
+type Analitycs = {
+  mambers: {
+    name: string;
+    node_id: string;
+    projects_names: {
+      url: string;
+      name: string;
+    }[];
+    avatar_url: string;
+    score: number;
+    stats: {
+      additions: number;
+      deletions: number;
+      commits: number;
+    };
+  }[];
+  since: string;
+  until: string;
+};
+
+type GithubContributorStats = z.infer<typeof gitHubContributorStatsSchema>;
+
+export function calculateTotals(
+  weeks: GithubContributorStats[number]['weeks'],
+) {
+  return weeks.reduce(
+    (acc, week) => {
+      const stats = {
+        additions: week.a + acc.stats.additions,
+        deletions: week.d + acc.stats.deletions,
+        commits: week.c + acc.stats.commits,
+      };
+      return {
+        score:
+          acc.score +
+          (stats.additions * 3 + stats.deletions * 2 + stats.commits),
+        stats,
+      };
+    },
+    { score: 0, stats: { additions: 0, deletions: 0, commits: 0 } },
+  );
```

Comment:
Thank you for raising those important points regarding date range filtering and weighing recent contributions. I appreciate the opportunity to address your concerns and provide further clarification.
Regarding the date range filtering, you are correct that the current implementation already includes the since and until parameters, which allow us to fetch data within a specific date range. This approach ensures that we only retrieve the necessary data, avoiding the need for additional filtering and potentially improving performance.
As for giving extra weight to recent contributions, I understand the potential value of this feature, as it could provide a more accurate and up-to-date representation of the data. However, I agree with your pragmatic approach of prioritizing the core functionality that will yield the most significant results.
In the spirit of the Pareto principle, which suggests that 80% of the results often come from 20% of the effort, it would be prudent to focus our efforts on delivering the essential features that address the primary use cases. Once we have established a solid foundation and achieved the desired 80% outcome, we can then consider enhancing the system with additional features like weighted contributions.
By concentrating on the critical functionality first, we can ensure a timely and effective delivery of the core product, while leaving room for future enhancements and optimizations. This iterative approach not only allows us to meet the immediate needs of our stakeholders but also provides flexibility to adapt and evolve the system as new requirements or opportunities arise.
Rest assured, the suggestion of giving extra weight to recent contributions has been noted, and we can revisit it during the subsequent phases of our project. At that point, we can conduct a thorough analysis of the potential benefits, trade-offs, and implementation considerations, ensuring that any additional features align with our overall architectural principles and deliver tangible value to our contributors in Maakaf.
I appreciate your understanding and support in prioritizing the essential work that will yield the most significant impact. Should you have any further concerns or suggestions, please feel free to raise them, and we can discuss them in detail.